### ISSUE TITLE
Warn about use of conditional null operator with an extension method that accepts null reference

### ISSUE THREAD

**Issue by mikenunan** [dotnet/roslyn#57298]
Example code in project targeting `net5.0`:

``` C#
using System;
using System.Collections.Generic;
using System.Linq;

#nullable enable

NullPropagationWithExtensionMethodExperiment(null);

static void NullPropagationWithExtensionMethodExperiment(IEnumerable<int>? integers)
{
    var result = integers?.SumIntegers(); // null-coalesce should be warned of here IMO
    Console.WriteLine(result?.Sum);
}

public record Result(int Sum);

public static class MyExtensions
{
    public static Result SumIntegers(this IEnumerable<int>? integers) => new(integers?.Sum() ?? 42);
}
```

As written, this will output a blank line to the console because the call to the `SumIntegers` extension method is bypassed by the conditional null operator on the line with the comment. If that `?.` is replaced by a regular `.` then `SumIntegers` is called and the value of 42 is returned and output. If an extension method accepts a null and potentially has specific handling for that case, then a warning is helpful.

I opened a discussion on this on SO here: https://stackoverflow.com/questions/69435088/possible-to-get-a-warning-for-use-of-conditional-null-operator-with-extension-me

While it's valid to say that the conditional null operator is just syntactic sugar, the use of an extension method disguises the situation so I think a warning is warranted, even if it is not enabled in the default rule set.

**Comment by mikenunan** [comment]
@CyrusNajmabadi my example was slightly contrived versus the prod code case, just to keep it short. In the real situation the extension method converted a DTO from an upstream system into another object, and the desire was to have that second object be non-null with null properties in the case when the incoming DTO was null. Ext methods for arithmetical operators on the second type (a representation of P&L in a financial system) need to do the same thing. Changing everything to regular extension methods broke the nice fluent syntax and completely destroyed the readability.

But either which way, if an extension method in null-aware code accepts a `WhateverType?` then the edge case is to call it with `?.` not `.`, surely?

**Comment by RikkiGibson** [comment]
It feels like there's an assumption that most null-accepting methods will basically do

```cs
static C2 M(this C1? c)
{
    if (c == null) { return null; }
    // ...
}
```

FWIW, this means that if the method is called in the form `c?.M()`, then a redundant null check is introduced.

**Comment by CyrusNajmabadi** [comment]
> In the real situation the extension method converted a DTO from an upstream system into another object, and the desire was to have that second object be non-null with null properties in the case when the incoming DTO was null. 

This itself feels like an antipattern to me.  In that case, i would write an analzyer specific to this API surface that checks for proper use.

**Comment by mikenunan** [comment]
> > My question is what is the benefit in that?
> 
> I mean... that is the benefit. :)
> 

Sorry I'm not following then, can you spell it out for me, if it's not performance or code bytes? (Neither of which were a concern to me anyway, I was just trying to understand the motivation against what I'm suggesting.)

What's left, avoidance of the extra character in the source? (To be completely clear, what I'm asking is why you would have your extension method take the nullable when all it does is return null-for-null, versus using the null-coalesce at the call site? This is a genuine question)

**Comment by CyrusNajmabadi** [comment]
> but you reject it because it falls outside project's internal needs from the language

That's not why i personally rejected it.  i'm rejecting it because i think things are working *properly* here :)  I've also suggested what i think is the *ideal* solution here. Namely, writing a custom analyzer that can encode the particulars of this specific design you have.

The point of me bringing up how we do things with our own code was simply to counter the idea that this was some sort of inappropriate pattern that would motivate *us* providing a warning here.  If this is highly valuable in many cases, but only not valuable for your particualr API, then an analyzer is *amazing* for this.

**Comment by CyrusNajmabadi** [comment]
> mistakes will be made, the more that the tooling can do to protect us against them, the better.

Agreed.  

However, we have to balance this against our thoughts on if things are 'mistakes' or not.  In this case, evaluating this situation i do *not* thnik this is a mistake.  I think this is *desirable*.  I agree that you tripped here, but we do not believe it is the responsibility to provide a warning for any particular location someone trips (esp. due to design decisions they made for their particular APIs).  In general, we are even very cautious about adding warnings.  We generally only add them when the following is true:

1. it is almost always a mistake in the code, and we regret that we ever allowed it.  indeed, we would prefer this to be an error, but we can't for back compat.

Warnings for things that are very reasonable for many codebases, but not for all, just leads to a very unpleasant scenario where tons of projects get warnings for things that are completely legit.  Imagine if each C# release started telling you there were problems for your code that you knew was completely fine.  It would make upgrades very unpleasant (and customers have told us this numerous times).

This is literally one of hte *primary reasons* we created analyzers in the first place.  We recognize that there are needs for domain/team-specific warnings for issues that arise for particular customers (or sets of customers).  These are appropraite when there are legitimate use cases out there and high likelihood that many customers are making use of this.  Customers who want to a harder line stance now can analyze their code with the same fidelity we perform analysis with.  They can state: yes, i know that this is ok for some people, but i don't think its ok for us.  So i will disallow.

**Comment by mavasani** [comment]
>  it doesn't seem innately objectionable to introduce some analyzer for detecting redundant or unnecessary null tests

@RikkiGibson Given that this code pattern indicates code smell rather than a certain bug that would qualify for a compiler warning, it seems fine to move it back to roslyn-analyzers and open it for community to contribute a CA analyzer for this.
